File: frontend/src/api/analytics.ts
import { buildQuery, request } from "./client";
import type { MetricKind, WeekOut } from "./types";

export interface MetricSummary {
  code: string;
  title: string;
  unit: string | null;
  direction: "higher_is_better" | "lower_is_better";
  kind: MetricKind;
  target: number;
  value: number | null;
  previous: number | null;
  delta: number | null;
  improved: boolean | null;
  reported: number;
  total: number;
  coverage: number | null;
  below_target: number;
}

export interface TrendPoint {
  week_id: number | null;
  label: string;
  starts_on: string;
  value: number | null;
  reported: number;
}

export interface ComparisonSeries {
  id: number;
  name: string;
  value: number | null;
  previous: number | null;
  delta: number | null;
  improved: boolean | null;
  reported: number;
  total: number;
  values: (number | null)[];
}

export interface AnalyticsOperator {
  user_id: number;
  full_name: string;
  group_id: number | null;
  group_name: string | null;
  value: number | null;
  previous: number | null;
  delta: number | null;
  improved: boolean | null;
  target_met: boolean | null;
  points: number | null;
  rank: number | null;
  missing_metrics: string[];
  values: Record<string, number | null>;
  trend: (number | null)[];
}

export interface AnalyticsOut {
  week: WeekOut | null;
  /** day, week or month: what the selected period and each trend point mean. */
  grain: AnalyticsGrain;
  period_label: string | null;
  period_from: string | null;
  period_to: string | null;
  metric_code: string | null;
  operator_count: number;
  operators_with_data: number;
  pending_requests: number;
  coins_awarded: number;
  metrics: MetricSummary[];
  trend: TrendPoint[];
  groups: ComparisonSeries[];
  comparisons: ComparisonSeries[];
  operators: AnalyticsOperator[];
  methodology: string;
}

export type AnalyticsGrain = "day" | "week" | "month";

export interface AnalyticsFilters {
  week_id?: number;
  grain?: AnalyticsGrain;
  date_from?: string;
  date_to?: string;
  group_id?: number;
  metric_code?: string;
  operator_ids?: string;
}

export const analytics = {
  overview: (filters: AnalyticsFilters, signal?: AbortSignal) =>
    request<AnalyticsOut>(`/api/v1/analytics/overview${buildQuery({ ...filters })}`, { signal }),
  summary: (filters: AnalyticsFilters, signal?: AbortSignal) =>
    request<AnalyticsOut>(`/api/v1/analytics/summary${buildQuery({ ...filters })}`, { signal }),
  mine: (filters: Pick<AnalyticsFilters, "week_id" | "metric_code"> = {}, signal?: AbortSignal) =>
    request<AnalyticsOut>(`/api/v1/analytics/me${buildQuery({ ...filters })}`, { signal }),
};

File: app/schemas/analytics.py
"""Analytics over recorded weekly values; absent observations remain null."""

from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field

from app.models.enums import MetricDirection, MetricKind
from app.schemas.rating import WeekOut


class MetricSummary(BaseModel):
    code: str
    title: str
    unit: str | None
    direction: MetricDirection
    kind: MetricKind
    target: float
    value: float | None
    previous: float | None
    delta: float | None
    improved: bool | None
    reported: int
    total: int
    coverage: float | None
    below_target: int


class TrendPoint(BaseModel):
    week_id: int | None
    label: str
    starts_on: date
    value: float | None
    reported: int


class ComparisonSeries(BaseModel):
    id: int
    name: str
    value: float | None
    previous: float | None
    delta: float | None
    improved: bool | None
    reported: int
    total: int
    values: list[float | None]


class AnalyticsOperator(BaseModel):
    user_id: int
    full_name: str
    group_id: int | None
    group_name: str | None
    value: float | None
    previous: float | None
    delta: float | None
    improved: bool | None
    target_met: bool | None
    points: float | None
    rank: int | None
    missing_metrics: list[str]
    values: dict[str, float | None]
    trend: list[float | None]


class AnalyticsOut(BaseModel):
    week: WeekOut | None
    #: day, week or month: what one point of the trend and the selected period mean.
    grain: str = "week"
    period_label: str | None = None
    period_from: date | None = None
    period_to: date | None = None
    metric_code: str | None
    operator_count: int
    operators_with_data: int
    pending_requests: int
    coins_awarded: int
    metrics: list[MetricSummary] = Field(default_factory=list)
    trend: list[TrendPoint] = Field(default_factory=list)
    groups: list[ComparisonSeries] = Field(default_factory=list)
    comparisons: list[ComparisonSeries] = Field(default_factory=list)
    operators: list[AnalyticsOperator] = Field(default_factory=list)
    methodology: str = (
        "Среднее по загруженным значениям действующих операторов. "
        "Группы и состав команды — текущие. Цели — действующие настройки. "
        "Пропуски не считаются нулями. Число отдельных проверок качества не хранится."
    )
